slither/utils/upgradeability.py

```diff
@@ -0,0 +1,463 @@
+    Contract,
+    Structure,
+    Enum,
+    SolidityVariableComposed,
+    SolidityVariable,
+    Function,
+)
+from slither.core.solidity_types import (
+    ElementaryType,
+    ArrayType,
+    MappingType,
+    UserDefinedType,
+)
+from slither.core.variables.local_variable import LocalVariable
+from slither.core.variables.local_variable_init_from_tuple import LocalVariableInitFromTuple
+from slither.core.variables.state_variable import StateVariable
+from slither.analyses.data_dependency.data_dependency import get_dependencies
+from slither.core.variables.variable import Variable
+from slither.core.expressions.literal import Literal
+from slither.core.expressions.identifier import Identifier
+from slither.core.expressions.call_expression import CallExpression
+from slither.core.expressions.assignment_operation import AssignmentOperation
+from slither.core.cfg.node import Node, NodeType
+from slither.slithir.operations import (
+    Assignment,
+    Index,
+    Member,
+    Length,
+    Binary,
+    Unary,
+    Condition,
+    NewArray,
+    NewStructure,
+    NewContract,
+    NewElementaryType,
+    SolidityCall,
+    Delete,
+    EventCall,
+    LibraryCall,
+    InternalDynamicCall,
+    HighLevelCall,
+    LowLevelCall,
+    TypeConversion,
+    Return,
+    Transfer,
+    Send,
+    Unpack,
+    InitArray,
+    InternalCall,
+)
+from slither.slithir.variables import (
+    TemporaryVariable,
+    TupleVariable,
+    Constant,
+    ReferenceVariable,
+)
+from slither.tools.read_storage.read_storage import SlotInfo, SlitherReadStorage
+
+
+# pylint: disable=too-many-locals
+def compare(v1: Contract, v2: Contract) -> dict:
+    """
+    Compares two versions of a contract. Most useful for upgradeable (logic) contracts,
+    but does not require that Contract.is_upgradeable returns true for either contract.
+
+    Args:
+        v1: Original version of (upgradeable) contract
+        v2: Updated version of (upgradeable) contract
+
+    Returns: dict {
+        "missing-vars-in-v2": list[Variable],
+        "new-variables": list[Variable],
+        "tainted-variables": list[Variable],
+        "new-functions": list[Function],
+        "modified-functions": list[Function],
+        "tainted-functions": list[Function]
+    }
+    """
+
+    order_vars1 = [v for v in v1.state_variables if not v.is_constant and not v.is_immutable]
+    order_vars2 = [v for v in v2.state_variables if not v.is_constant and not v.is_immutable]
+    func_sigs1 = [function.solidity_signature for function in v1.functions]
+    func_sigs2 = [function.solidity_signature for function in v2.functions]
+
+    results = {
+        "missing-vars-in-v2": [],
+        "new-variables": [],
+        "tainted-variables": [],
+        "new-functions": [],
+        "modified-functions": [],
+        "tainted-functions": [],
+    }
+
+    # Since this is not a detector, include any missing variables in the v2 contract
+    if len(order_vars2) < len(order_vars1):
```

Comment:
It looks like we are dupplicating existing logic. Ex: https://github.com/crytic/slither/blob/master/slither/tools/upgradeability/checks/variables_order.py#L51-L68

I think here we should:
- Create utils.upgradeability based on the upgradeability detectors 
- Use the utils in the upgradeability detectors

Reply:
My latest commit makes `get_missing_vars` its own utility function and uses it in the MissingVariable upgradeability detector.